dev/PushNotifications/PushNotificationsLongRunningTask/winmain.cpp

```diff
@@ -1,10 +1,111 @@
 ﻿// Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License. See LICENSE in the project root for license information.
 
-#include <windows.h>
 #include "pch.h"
 
+// include notifications constants file here
+
+#include <PushNotificationsLRP_h.h>
+
+// Temporarily disable C4324 because WRL generates a false (well, irrelevant) warning
+//   'Microsoft::WRL::Details::StaticStorage<Microsoft::WRL::Details::OutOfProcModuleBase<ModuleT>::GenericReleaseNotifier<T>,Microsoft::WRL::Details::StorageInstance::OutOfProcCallbackBuffer1,ModuleT>': structure was padded due to alignment specifier
+#pragma warning(push)
+#pragma warning(disable:4324)
+#include <wrl.h>
+#pragma warning(pop)
+
+#include "platform.h"
+#include "platformfactory.h"
+#include "externs.h"
+
+using namespace Microsoft::WRL;
+
+void SignalEvent()
+{
+    GetWinMainEvent().SetEvent();
+}
+
+wil::unique_threadpool_timer g_timer;
+
+void SetTimerForEvent()
+{
+    try
+    {
+        g_timer.reset(CreateThreadpoolTimer(
+            [](PTP_CALLBACK_INSTANCE, _Inout_ PVOID, _Inout_ PTP_TIMER)
+            {
+                SignalEvent();
+            },
+            reinterpret_cast<PVOID>(GetCurrentThreadId()),
+            nullptr));
+
+        THROW_LAST_ERROR_IF_NULL(g_timer);
+
+        // Negative times in SetThreadpoolTimer are relative. Allow 30 seconds to fire.
+        FILETIME dueTime{};
+        *reinterpret_cast<PLONGLONG>(&dueTime) = -static_cast<LONGLONG>(30000 * 10000);
+        SetThreadpoolTimer(g_timer.get(), &dueTime, 0, 0);
+    }
+    catch (...)
+    {
+        LOG_IF_FAILED(wil::ResultFromCaughtException());
+    }
+}
+
+// This function has to be called after we realized that we want to persist the LRP.
+void CancelTimerForEvent()
+{
+    g_timer.reset();
```

Comment:
What happens if 2 threads call reset here? Can this method race with SetTimerForEvent?

Reply:
No. We firstly set the timer, then we create the platform.